### ISSUE TITLE
"ORA-01795: maximum number of expressions in a list is 1000" when using inSet

### ISSUE THREAD

**Issue by ashleymercer** [slick/slick#1739]
Oracle has a hard limit of 1000 elements in an `IN` clause so, using the [canonical examples](http://slick.lightbend.com/doc/3.2.0/introduction.html#functional-relational-mapping), if I write something like:

```
val longList = Set(...)
coffees.filter(_.name inSet longList)
```

and `longList` has more than 1000 elements, then I get `ORA-01795: maximum number of expressions in a list is 1000`.

It's possible to fix this at the application layer, but it's fiddly since it has to be fixed individually in each place where I use `inSet`; it would be nice if there was a standard slick way of doing this. A couple of possible solutions I can think of:

- if we had a `DBIO.traverse` method (equivalent to something like `Future.traverse`) then it would at least be simpler to implement the batching in application code

- slick could automatically batch the list into blocks of 1000, run multiple queries inside the same transaction, and stitch them back together on the client side

- use a [global temporary table](https://docs.oracle.com/cd/B28359_01/server.111/b28310/tables003.htm#ADMIN11633) to store the values and then join. Not very portable since it requires a different temporary table for each data type (int, String etc) to be configured on the database

- use a [`VARRAY`-like](https://docs.oracle.com/cd/B28359_01/appdev.111/b28370/collections.htm#CHDEIDIC) object to pass the values down the database. Probably more portable since there's already a [`java.sql.Array` interface](http://docs.oracle.com/javase/6/docs/api/java/sql/Array.html) which the OJDBC library implements, but I think (?) this still requires some data type(s) to be declared on the database first

There are probably other solutions I haven't thought of.

**Comment by trevorsibanda** [comment]
@ashleymercer I don't quite think this is a bug, the limit is imposed by Oracle, not Slick. I'm not so sure on providing a work around for this. 

There is a S.O question [Why oracle IN clause has limit of 1000 only for static data?](https://stackoverflow.com/questions/19002792/why-oracle-in-clause-has-limit-of-1000-only-for-static-data) for this as well.

**Comment by ashleymercer** [comment]
That's true, but there are other cases where slick _does_ hide the complexity of individual database engines.

For example, Oracle didn't (until recently) have auto-incrementing columns and instead you had to use sequences. This, we can all agree, is a royal pain - so slick very kindly [maps its `AutoInc` to a slick-generated sequence](https://stackoverflow.com/a/27346979/13065).

To my mind, this `inSet` problem falls into exactly the same category - you absolutely could just pass responsibility back to the application developer, but it would be very _helpful_ (and make the end code much cleaner) if slick handled it for you.

**Comment by cvogt** [comment]
I agree with @ashleymercer. But this may be of lower priority than some other issues, unless demand rises. Let's keep it open.

**Comment by mirlabraga** [comment]
<https://mailtrack.io/> Sent with Mailtrack
<https://mailtrack.io/install?source=signature&lang=en&referral=mirlabraga@gmail.com&idSignature=22>

Mirla Braga

**Comment by cvogt** [comment]
@mirlabraga looks like you have something misconfigured there, posting messages here
